broadcast/broadcast.go

```diff
@@ -49,17 +56,49 @@ func NewBroadcaster(n int) *Broadcaster {
 	return &Broadcaster{capacity: n}
 }
 
-// Send broadcasts a message to the channel.
+// SendWithTimeout broadcasts a message to each listener's channel.
 // Sending on a closed channel causes a runtime panic.
-func (b *Broadcaster) Send(v interface{}) {
+// This method blocks for a duration of up to `timeout` on each channel.
+// Returns error(s) if it is unable to send on a given listener's channel within `timeout` duration.
+func (b *Broadcaster) SendWithTimeout(v interface{}, timeout time.Duration) error {
 	b.m.Lock()
 	defer b.m.Unlock()
 	if b.closed {
 		panic("broadcast: send after close")
 	}
-	for _, l := range b.listeners {
-		l <- v
+	var result *multierror.Error
+	for id, l := range b.listeners {
+		select {
+		case l <- v:
+			// Success!
+		case <-time.After(timeout):
+			err := fmt.Sprintf("unable to send to listener '%d'", id)
+			result = multierror.Append(result, errors.New(err))
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+// Send broadcasts a message to each listener's channel.
+// Sending on a closed channel causes a runtime panic.
+// This method is non-blocking, and will return errors if unable to send on a given listener's channel.
+func (b *Broadcaster) Send(v interface{}) error {
+	b.m.Lock()
+	defer b.m.Unlock()
+	if b.closed {
+		panic("broadcast: send after close")
+	}
+	var result *multierror.Error
+	for id, l := range b.listeners {
+		select {
+		case l <- v:
+			// Success!
+		default:
```

Comment:
I think this can't happen since the first `case` failing should mean the receiver channel is closed and thus the program will panic.

Reply:
Are you sure? I think this can happen if the buffer is full also right? This means it just won't block, and pass on to default?

Reply:
Oh yes, of course... got biased by the panic comment lines before hehe.
You're right if the buffer is full (or if it is unbuffered and no receiver blocked at it) will default.